service/service.go

```diff
@@ -20,75 +20,89 @@ import (
 	"context"
 	"fmt"
 	"net"
-	"os"
-	"os/signal"
-	"sync"
-	"syscall"
 
 	"github.com/plgd-dev/client-application/service/device"
 	"github.com/plgd-dev/client-application/service/grpc"
 	"github.com/plgd-dev/client-application/service/http"
+	"github.com/plgd-dev/hub/v2/pkg/fn"
 	"github.com/plgd-dev/hub/v2/pkg/fsnotify"
 	"github.com/plgd-dev/hub/v2/pkg/log"
+	"github.com/plgd-dev/hub/v2/pkg/service"
 	"go.opentelemetry.io/otel/trace"
 )
 
-type Service struct {
-	config         Config
-	ctx            context.Context
-	cancel         context.CancelFunc
-	logger         log.Logger
-	httpService    *http.Service
-	grpcService    *grpc.Service
-	deviceService  *device.Service
-	tracerProvider trace.TracerProvider
-	sigs           chan os.Signal
+const serviceName = "client-application"
+
+func newHttpService(ctx context.Context, config Config, clientApplicationServer *grpc.ClientApplicationServer, fileWatcher *fsnotify.Watcher, logger log.Logger, tracerProvider trace.TracerProvider) (*http.Service, error) {
+	httpService, err := http.New(ctx, serviceName, config.APIs.HTTP.Config, clientApplicationServer, fileWatcher, logger, tracerProvider)
+	if err != nil {
+		return nil, fmt.Errorf("cannot create http service: %w", err)
+	}
+	scheme := "https"
+	addr := getAddress(httpService.Address())
+	log.Infof("HTTP API available on %v://%s%v", scheme, addr, http.ApiV1)
+	if config.APIs.HTTP.UI.Enabled {
+		log.Infof("HTTP UI available on %v://%s", scheme, addr)
+	}
+	return httpService, err
 }
 
-const serviceName = "client-application"
+func newGrpcService(ctx context.Context, config Config, clientApplicationServer *grpc.ClientApplicationServer, fileWatcher *fsnotify.Watcher, logger log.Logger, tracerProvider trace.TracerProvider) (*grpc.Service, error) {
+	grpcService, err := grpc.New(ctx, serviceName, config.APIs.GRPC.Config, clientApplicationServer, fileWatcher, logger, tracerProvider)
+	if err != nil {
+		return nil, err
+	}
+	addr := getAddress(grpcService.Address())
+	insecure := ""
+	if !config.APIs.GRPC.Config.TLS.Enabled {
+		insecure = " (insecure)"
+	}
+	log.Infof("gRPC API available on %s%v", addr, insecure)
+	return grpcService, nil
+}
+
+func closeHttpOnError(err error, services []service.APIService) error {
+	if len(services) > 0 {
+		err = fmt.Errorf("cannot create grpc service: %w", err)
+		err2 := services[0].Close()
```

Comment:
Isn't `service[0]` always deviceService and not httpService (see line 86 below)?

Reply:
fixed